ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala

```diff
@@ -61,70 +63,71 @@ private[apiserver] final class LedgerTimeAwareCommandExecutor(
+
+          def failed = Future.successful(Left(ErrorCause.LedgerTime(maxRetries - retriesLeft)))
+          def success(c: CommandExecutionResult) = Future.successful(Right(c))
+          def retry(c: Commands) = {
+            metrics.daml.execution.retry.mark()
+            loop(c, submissionSeed, ledgerConfiguration, retriesLeft - 1)
+          }
+
+          contractStore
+            .lookupMaximumLedgerTimeAfterInterpretation(usedContractIds)
+            .transformWith {
+              case Success(MaximumLedgerTime.NotAvailable) =>
+                success(cer)
+
+              case Success(MaximumLedgerTime.Max(maxUsedTime))
+                  if maxUsedTime <= commands.commands.ledgerEffectiveTime =>
+                success(cer)
+
+              case Success(MaximumLedgerTime.Max(maxUsedTime)) if !cer.dependsOnLedgerTime =>
+                logger.debug(
+                  s"Advancing ledger effective time for the output from ${commands.commands.ledgerEffectiveTime} to $maxUsedTime"
+                )
+                success(advanceOutputTime(cer, maxUsedTime))
+
+              case Success(MaximumLedgerTime.Max(maxUsedTime)) =>
+                if (retriesLeft > 0) {
                   logger.debug(
                     s"Restarting the computation with new ledger effective time $maxUsedTime"
                   )
-                  val advancedCommands = advanceInputTime(commands, maxUsedTime)
-                  loop(advancedCommands, submissionSeed, ledgerConfiguration, retriesLeft - 1)
+                  retry(advanceInputTime(commands, maxUsedTime))
                 } else {
-                  Future.successful(Left(ErrorCause.LedgerTime(maxRetries)))
+                  failed
                 }
-              }
-              .recoverWith {
-                case MissingContracts(contracts) =>
-                  if (retriesLeft > 0) {
-                    metrics.daml.execution.retry.mark()
-                    logger.info(
-                      s"Some input contracts could not be found. Restarting the computation. Missing contracts: ${contracts
-                        .mkString("[", ", ", "]")}"
-                    )
-                    loop(commands, submissionSeed, ledgerConfiguration, retriesLeft - 1)
-                  } else {
-                    logger.info(
-                      s"Lookup of maximum ledger time failed after ${maxRetries - retriesLeft}. Used contracts: ${usedContractIds
-                        .mkString("[", ", ", "]")}."
-                    )
-                    Future.successful(Left(ErrorCause.LedgerTime(maxRetries)))
-                  }
-
-                // An error while looking up the maximum ledger time for the used contracts. The nature of this error is not known.
-                // Not retrying automatically. All other automatically retryable cases are covered by the logic above.
-                case error =>
+
+              case Success(Archived(contracts)) =>
```

Comment:
```suggestion
              case Success(MaximumLedgerTime .Archived(contracts)) =>
```

to make it consistent with the other cases